CUDAGraph Trees - Warn on dealloc

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97378
* #97440
* #97176
* #96775
* #97419
* __->__ #97171



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

eellison committed:
Warn on dealloc

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44228370).

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

ezyang commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -1345,9 +1373,14 @@ def try_end_curr_warmup(self):
                 torch._C._free_And_Remove_DeleterFn(t())
+                warnings.warn(
+                    f"CUDA Graphs is Deallocating Tensor Output from {stack_trace.strip()}. "
+                    "Deallocate output tensor no longer in use or copy output."
```
This is not a good warning. Make it better.

eellison replied:
What would you suggest ?

ezyang replied:
CUDAGraphTree triggered deallocation of tensor {id(t)} due to REASON, even though this tensor is still live. Subsequent use of this tensor may return garbage result.

Also, if we have access to the old Tensor objects, I wonder if we can modify them to make them error if you try to read. This would be much much better than this warning (which is not very actionable).

eellison replied:
Tensors arent around here, just storages. I agree making storage access fail would be good if it's possible.

ezyang replied:
Well, all I am really asking for here is a warning that is more clear. Imagine someone is debugging a SEV where there is an NE explosion and they're looking through the warnings and they see this warning. Will they be able to figure it out from this warning text? What can help them figure it out?

eellison replied:
The stack trace and the suggested message (copy the output) is a good start IMO, and gives an actionable message. 

I also think there are other changes to be made that should make the whole thing safer.

ezyang replied:
I haven't the foggiest how I would copy the output. Let's take your test case as an example:

```
        def test_warnings_on_dealloc(self):
            @torch.compile()
            def foo(x):
                return x * x * x

            inp = torch.rand([4], device="cuda")
            out = foo(inp)
            warnings.resetwarnings()
            with warnings.catch_warnings(record=True) as w:
                foo(inp)

            self.assertTrue(len(w) == 1)
            self.assertTrue("x * x * x" in str(w[0]))
```

According to the test, the warning gives you a stack trace that identifies "return x * x * x". So can I fix the problem by saying "return (x * x * x).clone()"? This seems unlikely to help, as the clone will get shoved in the CUDA graph.

eellison replied:
Copy the output of `torch.compile` is more clear. `tensor {id(t)}` is a temporary value because aot autograd will detach, so I'm not sure what value it has to user

eellison replied:
```
            @torch.compile()
            def foo(x):
                return x * x * x

            inp = torch.rand([4], device="cuda")
            out = foo(inp).clone()
 ```

ezyang replied:
Don't explain this to me, explain it to the user, in the warning!

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison committed:
Update on "CUDAGraph Trees - Warn on dealloc"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D44228370](https://our.internmc.facebook.com/intern/diff/D44228370)

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
